influxdb3_server/src/builder.rs

```diff
@@ -57,65 +60,94 @@ pub struct WithPersister<P>(Arc<P>);
             write_buffer: WithWriteBuf(wb),
             query_executor: self.query_executor,
             persister: self.persister,
+            listener: self.listener,
             authorizer: self.authorizer,
         }
     }
 }
 
-impl<W, P, T> ServerBuilder<W, NoQueryExec, P, T> {
-    pub fn query_executor<Q>(self, qe: Arc<Q>) -> ServerBuilder<W, WithQueryExec<Q>, P, T> {
+impl<W, P, T, L> ServerBuilder<W, NoQueryExec, P, T, L> {
+    pub fn query_executor<Q>(self, qe: Arc<Q>) -> ServerBuilder<W, WithQueryExec<Q>, P, T, L> {
         ServerBuilder {
             common_state: self.common_state,
             time_provider: self.time_provider,
             max_request_size: self.max_request_size,
             write_buffer: self.write_buffer,
             query_executor: WithQueryExec(qe),
             persister: self.persister,
+            listener: self.listener,
             authorizer: self.authorizer,
         }
     }
 }
 
-impl<W, Q, T> ServerBuilder<W, Q, NoPersister, T> {
-    pub fn persister<P>(self, p: Arc<P>) -> ServerBuilder<W, Q, WithPersister<P>, T> {
+impl<W, Q, T, L> ServerBuilder<W, Q, NoPersister, T, L> {
+    pub fn persister<P>(self, p: Arc<P>) -> ServerBuilder<W, Q, WithPersister<P>, T, L> {
         ServerBuilder {
             common_state: self.common_state,
             time_provider: self.time_provider,
             max_request_size: self.max_request_size,
             write_buffer: self.write_buffer,
             query_executor: self.query_executor,
             persister: WithPersister(p),
+            listener: self.listener,
             authorizer: self.authorizer,
         }
     }
 }
 
-impl<W, Q, P> ServerBuilder<W, Q, P, NoTimeProvider> {
-    pub fn time_provider<T>(self, tp: Arc<T>) -> ServerBuilder<W, Q, P, WithTimeProvider<T>> {
+impl<W, Q, P, L> ServerBuilder<W, Q, P, NoTimeProvider, L> {
+    pub fn time_provider<T>(self, tp: Arc<T>) -> ServerBuilder<W, Q, P, WithTimeProvider<T>, L> {
         ServerBuilder {
             common_state: self.common_state,
             time_provider: WithTimeProvider(tp),
             max_request_size: self.max_request_size,
             write_buffer: self.write_buffer,
             query_executor: self.query_executor,
             persister: self.persister,
+            listener: self.listener,
+            authorizer: self.authorizer,
+        }
+    }
+}
+
+impl<W, Q, P, T> ServerBuilder<W, Q, P, T, NoListener> {
+    pub fn tcp_listener(self, listener: TcpListener) -> ServerBuilder<W, Q, P, T, WithListener> {
+        ServerBuilder {
+            common_state: self.common_state,
+            time_provider: self.time_provider,
+            max_request_size: self.max_request_size,
+            write_buffer: self.write_buffer,
+            query_executor: self.query_executor,
+            persister: self.persister,
+            listener: WithListener(listener),
             authorizer: self.authorizer,
         }
```

Comment:
I think this can just be

`ServerBuilder { listener: WithListener(listener), ..self }`

Reply:
Same with the other methods like this. Should reduce some line noise

Reply:
This is unfortunately part of the pain and suffering of this excessively boiler-plate-y builder pattern. Because the return type is different than the `Self` type, the compiler doesn't allow this 😔:
```
error[E0308]: mismatched types
  --> influxdb3_server/src/builder.rs:72:15
   |
72 |             ..self
   |               ^^^^ expected `ServerBuilder<WithWriteBuf<...>, ..., ..., ..., ...>`, found
 `ServerBuilder<NoWriteBuf, ..., ..., ..., ...>`
   |
   = note: expected struct `ServerBuilder<WithWriteBuf<W>, _, _, _, _>`      (return type)
              found struct `ServerBuilder<NoWriteBuf, _, _, _, _>`           (Self type)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `influxdb3_server` (lib) due to 1 previous error
```